torch/csrc/autograd/engine.cpp

```diff
@@ -824,9 +824,22 @@ void Engine::evaluate_function(
   }
 }
 
-/* Computes the number of dependencies for each function which requires grad */
-auto Engine::compute_dependencies(Node* root, GraphTask& task) -> void {
-  // Just to make sure that they will never be added to the queue again
+inline static uint64_t compute_min_sequence_nr(const edge_list& outputs) {
+  // Computes the mininum sequence number among all the outputs and applies
+  // a small constant discount to take into account the lazy creation of AccumulateGrad
+  auto min_seq_nr = std::numeric_limits<uint64_t>::max();
+  for (auto & output_edge : outputs) {
+    Node *output = output_edge.function.get();
+    auto seq_nr = output->actual_sequence_nr();
+    min_seq_nr = (min_seq_nr < seq_nr) ? min_seq_nr : seq_nr;
+  }
+  // AccumulateGrad may have an actual_sequence_nr greater than its parent
+  const size_t discount = 10;
+  return min_seq_nr > discount ? min_seq_nr - discount : 0;
+}
```

Comment:
Is this the right approach? If so, how do we determine this constnat?

Reply:
The discount was necessary because we first create the Node then call collect_next_edges, which calls gradient_edge, so AccumulateGrad is guaranteed to have a sequence_nr greater than its parent. If we don't have discount, then the first parent of AccumulateGrad may have a seq_nr lower than min_seq_nr which is the seq_nr of the outputs. This would cause us to skip recursing when we reach AccumulateGrad's parent.

Can you elaborate what you mean by "AccumulateGrad Node is cached on the Tensor right"?

What do you mean by "AccumulateGrad Node is cached on the Tensor".

Reply:
I was thinking of https://github.com/pytorch/pytorch/blob/5003d417d41346aed628aa802eafa8e8bd9b79cc/torch/csrc/autograd/variable.h#L195
In particular, if you do a simple training loop where the previous iteration graph is freed after you have done the forward of the new iteration, then the AccumulateGrad Node will actually stay alive for the whole epoch:

```python
import torch

a = torch.rand(10, requires_grad=True)

for i in range(10):
    out = (a + 1) # Comment this line out to get different allocations
    # Print metadata here as it never changes for a given Node. Contrary to the
    # python binding object that is recrerated as it is not owned from c++
    print(id((a+1).grad_fn.next_functions[0][0].metadata))

    # Unrelated stuff to reduce chance of metadata dict being allocated at the
    # exact same place by python for different Nodes
    b = {}
```